MockedRethinkDB.js

```diff
@@ -0,0 +1,334 @@
+    });
+
+    test.beforeEach("generate temporary RethinkDB storage", t => {
+      return t.context[opts.internalContext].storage();
+    });
+
+    test.beforeEach("spawn local RethinkDB server", t => {
+      return t.context[opts.internalContext].spawn();
+    });
+
+    test.beforeEach("hack the 'rethinkdb' package to use local server", t => {
+      return t.context[opts.internalContext].hackModule();
+    });
+
+    if(opts.initialData) {
+      test.beforeEach("populate the RethinkDB server with data", t => {
+        return t.context[opts.internalContext].importData();
+      });
+    }
+
+    test.afterEach.always("shutdown 'rethinkdb' server", t => {
+      return t.context[opts.internalContext].shutdown();
+    });
+
+    test.afterEach.always("cleanup temporary RethinkDB storage", t => {
+      return t.context[opts.internalContext].removeStorage();
+    });
+  }
+
+  /**
+   * @param {Object} opts common options to mocking methods.
+  */
+  constructor(opts) {
+    this.opts = opts;
+  }
+
+  /**
+   * The port to assign to the spawned RethinkDB process.
+   * @return {Object} `{ offset, port }` the offset and port to use.
+   * @depreciated should not be used in anything except spawning RethinkDB, so spawn can try different ports if one is
+   *   in use.
+   * @todo If passed a port (e.g. port in use and we need another), select the next port to try.
+  */
+  port() {
+    if(this.opts.port && this.opts.offset) {
+      return {
+        offset: this.opts.offset,
+        port: this.opts.port,
+      };
+    }
+    const maxOffset = 65535 - 28015;
+    const pid = process.pid;
+    const offset = pid - (Math.floor(pid / maxOffset) * maxOffset);
+    return { offset, port: 28015 + offset };
+  }
+
+  /**
+   * Creates a temporary storage directory to contain the data for the mock database.
+   * @return {Promise<String>} resolves with the path for the storage directory once created.
+  */
+  storage() {
+    if(this._storage) { return this._storage; }
+    return this._storage = tmp
+      .dir()
+      .then(o => {
+        this._tmpDirData = o;
+        return o.path;
+      });
+  }
+
+  /**
+   * Spawn a new RethinkDB server instance, and wait for it to initialize.
+   * @return {Promise<Number>} resolves with the server's listening port when the process has started.
+   * @todo Retry with different ports if RethinkDB fails because of a taken port.
+  */
+  spawn() {
+    if(this._spawn) { return this._spawn; }
+    return this._spawn = Promise
+      .all([
+        this.port(),
+        this.storage(),
+      ])
+      .then(([ { offset }, storage ]) => {
+        this.spawned = spawn('rethinkdb', ['-o', `${offset}`, '-d', `${storage}`]);
+        return this.bootWait(this.spawned).then(() => port);
+      });
+  }
+
+  /**
+   * Hack the `require("rethinkdb")` module to use the mocked server.
+   * @return {Promise<RethinkDB>} resolves with the RethinkDB module after it's been modified
+  */
+  hackModule() {
```

Comment:
Might be better to name this something like `setRethinkDefaultPort` rather than `hackModule`. `hackModule`, although descriptive of what it's doing, doesn't describe the side effects.